etna/transforms/math/lags.py

```diff
@@ -96,3 +100,199 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
     def get_regressors_info(self) -> List[str]:
         """Return the list with regressors created by the transform."""
         return [self._get_column_name(lag) for lag in self.lags]
+
+
+class ExogShiftTransform(IrreversibleTransform, FutureMixin):
+    """Shifts exogenous variables from a given dataframe."""
+
+    def __init__(self, lag: Union[int, Literal["auto"]], horizon: Optional[int] = None):
+        """Create instance of ExogShiftTransform.
+
+        Parameters
+        ----------
+        lag:
+            value for shift estimation
+
+            * if set to `int` all exogenous variables will be shifted `lag` steps forward;
+
+            * if set to `auto` minimal shift will be estimated for each variable based on
+              the prediction horizon and available timeline
+
+        horizon:
+            prediction horizon. Mandatory when set to `lag="auto"`, ignored otherwise
+        """
+        super().__init__(required_features="all")
+
+        self.lag: Optional[int] = None
+        self.horizon: Optional[int] = None
+        self._auto = False
+
+        self._created_regressors: Optional[List[str]] = None
+        self._exog_shifts: Optional[Dict[str, int]] = None
+        self._exog_last_date: Optional[Dict[str, pd.Timestamp]] = None
+        self._filter_out_columns = {"target"}
+
+        if isinstance(lag, int):
+            if lag <= 0:
+                raise ValueError(f"{self.__class__.__name__} works only with positive lags values, {lag} given")
+            self.lag = lag
+
+        else:
+            if horizon is None:
+                raise ValueError("Value of `horizon` should be specified when using `auto`!")
+
+            if horizon < 1:
+                raise ValueError(f"{self.__class__.__name__} works only with positive horizon values, {horizon} given")
+
+            self.horizon = horizon
+            self._auto = True
+
+    def _save_exog_last_date(self, df_exog: pd.DataFrame):
+        """Save last available date of each exogenous variable."""
+        exog_names = set(df_exog.columns.get_level_values("feature"))
+
+        self._exog_last_date = dict()
+        for name in exog_names:
+            feature = df_exog.loc[:, pd.IndexSlice[:, name]]
+
+            na_mask = pd.isna(feature).any(axis=1)
+            last_date = feature.index[~na_mask].max()
+
+            self._exog_last_date[name] = last_date
+
+    def fit(self, ts: TSDataset) -> "ExogShiftTransform":
+        """Fit the transform.
+
+        Parameters
+        ----------
+        ts:
+            Dataset to fit the transform on.
+
+        Returns
+        -------
+        :
+            The fitted transform instance.
+        """
+        df_exog = ts.df_exog
+        if df_exog is not None and isinstance(df_exog, pd.DataFrame):
```

Comment:
Why do we need `isinstance` check here?

Reply:
Can't we put all the necessary logic of checking `isinstance` inside of `_save_exog_last_date`? I think we should also set `self._exog_last_date = {}` if there are no exogs.

Reply:
This check was needed for the base class. Now it seems irrelevant.